### PR TITLE
feature/INTHAD-147-GPS-fusion

### PR DESCRIPTION
**Pull request by shanmukhananda**
PR from `feature/INTHAD-147-GPS-fusion` to `feature/use-out-of-source-g2o`

Contains changes for using position from external source as prior edges.
DO-NOT REVIEW `ORB_SLAM2/experiments` its not part of slam app/lib.
`data/` folder contain report from testing hence PR has lot of files. Its recommended to review using comparison tool.

### CODE REVIEW

**Comment by shanmukhananda** [comment]
@paul-michalik  I have changed the base of GPS Fusion branch to `master`, building the experiments is now a cmake option `BUILD_EXPERIMENTS`

File: .vscode/settings.json [hunk 1]
```diff
@@ -78,6 +78,9 @@
         "strstream": "cpp",
         "complex": "cpp",
         "optional": "cpp",
-        "typeindex": "cpp"
+        "typeindex": "cpp",
+        "csignal": "cpp",
+        "codecvt": "cpp",
+        "future": "cpp"
```
Comment (paul-michalik) [inline .vscode/settings.json]: Why do you need this? Never seen anybody doing it. IntelliSense?

File: experiments/cpp/geoconv2/apps/app-oxts2local.cpp [hunk 2]
```diff
@@ -27,24 +27,22 @@ using rotation_t = std::tuple<double, double, double>;
 using point3d_t = std::tuple<double, double, double>;
 using sensor_data_t = std::tuple<geo_t, rotation_t>;
 
-void check_existence(const std::initializer_list<fs::path>& paths)
-{
+void check_existence(const std::initializer_list<fs::path>& paths) {
```
Comment (paul-michalik) [inline experiments/cpp/geoconv2/apps/app-oxts2local.cpp]: Just a side note: You'd rarely find usage explicit usage of `initializer_list` except of constructor. We prefer a generic implementation with a `template<class TRange> void foo(TRange&&);` or the traditional `template<class TIter> void foo(TIter b_, TIter e_);`

File: experiments/cpp/geoconv2/apps/app-oxts2local.cpp [hunk 3]
```diff
@@ -53,8 +51,7 @@ std::vector<fs::path> get_files(const fs::path& folder, const std::string& ext)
     return files;
 }
 
-std::vector<std::string> split(const std::string& s, const char delim)
-{
+std::vector<std::string> split(const std::string& s, const char delim) {
```
Comment (paul-michalik) [inline experiments/cpp/geoconv2/apps/app-oxts2local.cpp]: Are you aware of [Boost String Algorithms](https://www.boost.org/doc/libs/1_67_0/doc/html/string_algo.html)?

**Comment by paul-michalik** [comment]
It took me hours to make everything (incl. unit tests) run and behave in all environments properly... Visual Studio (via CMake Open Folder), via `configure.bat` and building on Ubuntu via Docker. Not sure if it runs on Ubuntu though since the `g2o_slamext.so` is not in path... Didn't push it yet, still some discrepancies with current version. We urgently need to set up a better process for verifying what needs to be runnable and when.

**Comment by paul-michalik** [comment]
You say I should not review the experiments... But may I make a proposal? Let us rewrite the experiments as detectable and interactively executable unit tests. You can grab either Boost.Test or GTest. (1) We obtain a reproducible environment for experiments (as we do for tests)  (2) We get the benefits of interactive or scripted experiment execution e.g. via test runner in Visual Studio etc.

**Comment by paul-michalik** [comment]
See #89! :-)

**Comment by shanmukhananda** [comment]
`experiments` contains apps which I used to learn certain things like calculating covariance using opencv APIs, compare different implementation of coordiante conversions. It also houses some silly apps which converts points3d to oxts files etc.
All these apps are just non-sense for an expert, hence I asked not to review these. 

I also disabled compiling of experiments in build script. `experiments` are created with a thought "_Lets try and see what happens_". Adding unit-tests are definitely good, which should be handled in seperate individual tasks.

If external position fusion implementation is only required, then I can prepare another branch which contains bear minimum changes like,
- `pos_adder.h`
- optimizer changes to use `add_pos_prior` and `add_pos_displacement`
- `ext_g2o`
- deleted `experiments`
- `Examples\Monocular\corrected_scale`  KITTI config files (Viewer changes)

This way its easy review and merge to master.

**Comment by paul-michalik** [comment]
No it's fine, leave it as it is. I'd rather create a new branch on which either experiments are formulated as unit tests with unified conventions (aka macs).

**Comment by paul-michalik** [comment]
I am adding support for debug builds  as the last thing. Then we should merge.

**Comment by shanmukhananda** [comment]
Hate to ruin the joy!.

I see linking is not happening correctly after using [this](https://github.com/paul-michalik/ORB_SLAM2/blob/fe962b3227c0a345509b30e91dff0cbee448b17b/ext_g2o/CMakeLists.txt#L21).

Debug and Release libraries are mixed during linking.

I tried printing cmake target properties after using [this](https://github.com/paul-michalik/ORB_SLAM2/blob/fe962b3227c0a345509b30e91dff0cbee448b17b/ext_g2o/CMakeLists.txt#L21). The result is expected, DEBUG config includes all `_d.so` and in NOT-DEBUG config there are no `_d.so`

```
$<$<CONFIG:DEBUG>:/usr/local/lib/libg2o_core_d.so>/usr/local/lib/libg2o_solver_eigen_d.so/usr/local/lib/libg2o_stuff_d.so/usr/local/lib/libg2o_types_data_d.so/usr/local/lib/libg2o_types_icp_d.so/usr/local/lib/libg2o_types_sba_d.so/usr/local/lib/libg2o_types_sclam2d_d.so/usr/local/lib/libg2o_types_sim3_d.so/usr/local/lib/libg2o_types_slam2d_d.so/usr/local/lib/libg2o_types_slam3d_d.so$<$<NOT:$<CONFIG:DEBUG>>:/usr/local/lib/libg2o_core.so>/usr/local/lib/libg2o_solver_eigen.so/usr/local/lib/libg2o_stuff.so/usr/local/lib/libg2o_types_data.so/usr/local/lib/libg2o_types_icp.so/usr/local/lib/libg2o_types_sba.so/usr/local/lib/libg2o_types_sclam2d.so/usr/local/lib/libg2o_types_sim3.so/usr/local/lib/libg2o_types_slam2d.so/usr/local/lib/libg2o_types_slam3d.so
```

But the output of `ldd` shows different data. `Release` config has `_d.so`
```
libg2o_slamext.so => /mnt/hgfs/D/git/ORB_SLAM2/products/cmake.make.linux.x86_64.gcc.7.Release/Release/lib/libg2o_slamext.so (0x00007f1c6f357000)
        libGL.so.1 => /usr/lib/x86_64-linux-gnu/libGL.so.1 (0x00007f1c6f0cb000)
        libg2o_stuff_d.so => /usr/local/lib/libg2o_stuff_d.so (0x00007f1c6ee7a000)
        libg2o_types_sba_d.so => /usr/local/lib/libg2o_types_sba_d.so (0x00007f1c6e3cb000)
        libg2o_types_sim3_d.so => /usr/local/lib/libg2o_types_sim3_d.so (0x00007f1c6dd9e000)
        libg2o_core.so => /usr/local/lib/libg2o_core.so (0x00007f1c6db3c000)
```
Apps crashes in linux due to this mix up. In Linux we can link Release mode .so to debug mode app, the other way(i.e. linking debug mode .so to release mode app) is not allowed.

Proposal:
Add `SET(G2O_LIBS ${G2O_LIBS_RELEASE} ${G2O_LIBS_DEBUG})` in FindG2O.cmake and use `target_link_libraries` without `debug`, `optimized` parameters

**Pull request merged**